flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/ShuffleOperator.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink.shuffle;
+
+import java.io.Serializable;
+import java.util.Map;
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.api.common.state.ListState;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.functions.KeySelector;
+import org.apache.flink.api.java.typeutils.MapTypeInfo;
+import org.apache.flink.runtime.operators.coordination.OperatorEvent;
+import org.apache.flink.runtime.operators.coordination.OperatorEventGateway;
+import org.apache.flink.runtime.operators.coordination.OperatorEventHandler;
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.runtime.state.StateSnapshotContext;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+/**
+ * Shuffle operator can help to improve data clustering based on the key.
+ *
+ * <p>It collects the data statistics information, sends to coordinator and gets the global data
+ * distribution weight from coordinator. Then it will ingest the weight into data stream(wrap by a
+ * class{@link ShuffleRecordWrapper}) and send to partitioner.
+ */
+@Internal
+public class ShuffleOperator<T, K extends Serializable>
+    extends AbstractStreamOperator<ShuffleRecordWrapper<T, K>>
+    implements OneInputStreamOperator<T, ShuffleRecordWrapper<T, K>>, OperatorEventHandler {
+
+  private static final long serialVersionUID = 1L;
+
+  private final KeySelector<T, K> keySelector;
+  // the type of the key to collect data statistics
+  private final TypeInformation<K> keyType;
+  private final OperatorEventGateway operatorEventGateway;
+  // key is generated by applying KeySelector to record
+  // value is the times key occurs
+  private transient Map<K, Long> localDataStatisticsMap;
```

Comment:
For high cardinality cases we might want to store the statistics in some `digest` implementation, like [t-digest](https://github.com/tdunning/t-digest) or [q-digest](https://github.com/addthis/stream-lib/blob/master/src/main/java/com/clearspring/analytics/stream/quantile/QDigest.java).

Do not need any specific change ATM, but we might want to keep the road open for this .

Reply:
high-cardinality is out of the scope of initial impl. But I see @pvary 's point regarding compatibility for future support. For that, we may want to introduce a `TrafficStatistics` interface. Currently, we can provide a `SimpleCountStatistics` impl.  We can implement a `TrafficStatisticsSerializer`. The serializer can be expanded in the future to handle high-cardinality statistics like `TDigestStatistics` e.g..

Reply:
nit: variable name can be simpler as `localStatistics`. `Map` suffix is not necessary.

Reply:
I think the data statistics interface will be very straightforward.
I would expect something like this:
```
public interface DataStatistic<K> {
    long size();
    void put(K key);
    void merge(DataStatistic<K> other);
}
```

I do not think that the ShuffleOperator needs to know anything about the statistics and the partitioners used.
My feeling is that the ShuffleOperator is just responsible to:
- Collect statistics
- Distribute the statistics to the Partitioners

The `Partitioners` are responsible to decide on the distribution based on the collected `DataStatistics`. Here, I am not sure we can separate the `Partitioner` from the actual strategy (Bin Pack, or Range) used, since fetching a distribution based on a key (Iceberg partition based - Map), or fetching a distribution based on a range list (Iceberg sorted tables - TreeSet) has different performance on the hot code path, so we could not fall back to the more general solution (TreeSet) for performance reasons.

I still feel that having an interface defined for `DataStatistics` would be a good direction even at this stage.

We can implement the `MapStatistics` easily like this:
```
public class MapStatistic<K> implements DataStatistic<K> {
    private Map<K, Long> data = Maps.newHashMap();

    @Override
    public long size() {
        return data.size();
    }

    @Override
    public void put(K key) {
        data.put(key, data.getOrDefault(key, 0L) + 1);
    }

    @Override
    public void merge(DataStatistic<K> other) {
        Preconditions.checkArgument(other instanceof MapStatistic, "Can not merge this type of statistics: " + other);
        MapStatistic mapStatistic = (MapStatistic<K>) other;
        Map<K, Long> otherData = mapStatistic.data;
        otherData.forEach((key, count) -> data.put(key, data.getOrDefault(key, 0L) + count));
    }
}
```

WTYT?

Reply:
Also, I am not sure that we want to seamlessly switch between different implementations. @stevenzwu's table describes the following possibilities in the doc where the strategy is based on the table definition and not on the statistics:
```
Partitioned \ Sorted | Unsorted           | Sorted
---------------------| ------------------ | ----------------------------
Unpartitioned        | None               | Range
---------------------| ------------------ | ----------------------------
Partitioned          | BinPacking / Hash  | BinPacking if SortOrder only contains partition columns
                     |                    | Range if SortOder contains non-partition columns
```

I think we can we can reasonably expect that we have enough memory to keep the statistics in a map if BinPacking is used. As for Range partitioning strategy, I would expect that we need to use some kind of Digest statistics. I think it would be a very rare corner case when the possible values are fluctuating so much that we need to change dynamically between the Digest and the Map based statistics.

In the meantime we might want to provide a `DataStatisticsFactory` for the `ShuffleOperator` to generate the empty statistics, like this:
```
public interface DataStatisticFactory<K> {
    DataStatistic<K> newDataStatistic();
    TypeInformation type();
}
```

Reply:
I like @pvary's proposal on the `DataStatistics` interface.

> As for Range partitioning strategy, I would expect that we need to use some kind of Digest statistics.

I am not sure about this. I can see range partition for non-partition columns with cardinality at hundreds or thousands, e.g., country code (string), event type (string).

Reply:
It sounds good to me to add `DataStatistics` interface. 
For the usage of `DataStatisticsFactory `, I think it will take table/mode configuration to decide which `DataStatistics` to generate. For now,  function `newDataStatistic/createDataStatistic` will always return `MapDataStatistic` since this is the only supported type. 
I will make the modification in the latest commit.

Reply:
Thanks for the change @yegangy0718!
Looks good to me